File: backend/app/api/src/main/kotlin/com/wallet/clover/api/controller/AdminController.kt
package com.wallet.clover.api.controller

import com.wallet.clover.api.common.CommonResponse
import com.wallet.clover.api.service.DataInitializationService
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

@RestController
@RequestMapping("/api/v1/admin")
class AdminController(
    private val dataInitializationService: DataInitializationService
) {
    // Define a safe scope for background tasks
    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    @PostMapping("/init/history")
    suspend fun initHistory(
        @RequestParam(required = false, defaultValue = "1") start: Int,
        @RequestParam(required = false) end: Int?
    ): CommonResponse<String> {
        // Run in background to avoid timeout
        serviceScope.launch {
            dataInitializationService.initializeWinningInfo(start, end)
        }
        return CommonResponse.success("History initialization started in background")
    }

    @PostMapping("/init/spots")
    suspend fun initSpots(
        @RequestParam(required = false, defaultValue = "1") start: Int,
        @RequestParam(required = false) end: Int?
    ): CommonResponse<String> {
        // Run in background
        serviceScope.launch {
            dataInitializationService.initializeWinningStores(start, end)
        }
        return CommonResponse.success("Store initialization started in background")
    }
}

File: apps/backend/src/admin/admin.controller.ts
import {
  Controller,
  Post,
  Query,
  ParseIntPipe,
  UseGuards,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { AdminService } from './admin.service';

@UseGuards(AuthGuard('jwt'))
@Controller('admin')
export class AdminController {
  constructor(private readonly adminService: AdminService) {}

  @Post('init/history')
  async initHistory(
    @Query('start', new ParseIntPipe({ optional: true })) start?: number,
    @Query('end', new ParseIntPipe({ optional: true })) end?: number,
  ) {
    const message = await this.adminService.initializeWinningInfo(
      start ?? 1,
      end,
    );
    return { message };
  }

  @Post('init/spots')
  async initSpots(
    @Query('start', new ParseIntPipe({ optional: true })) start?: number,
    @Query('end', new ParseIntPipe({ optional: true })) end?: number,
  ) {
    const message = await this.adminService.initializeWinningStores(
      start ?? 1,
      end,
    );
    return { message };
  }
}
